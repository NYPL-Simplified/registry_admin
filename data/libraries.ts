/* eslint-disable camelcase */

export interface LibrariesData {
  uuid: string;
  basic_info: {
    name: string;
    description: string;
    short_name: string;
    number_of_patrons: string;
  };
  urls_and_contact: {
    authentication_url: string;
    contact_email: string;
    copyright_email: string;
    help_email: string;
    opds_url: string;
    web_url: string;
  };
  areas: {
    focus: string[];
  };
  stages: {
    library_stage: 'testing' | 'production' | 'canceled';
    registry_stage: 'testing' | 'production' | 'canceled';
  };
}

const libraries: LibrariesData[] = [
  {
    uuid: '1',
    basic_info: {
      name: 'Acton Public Library',
      description: 'Serving Old Saybrook, CT',
      short_name: 'CTACTN',
      number_of_patrons: '9',
    },
    urls_and_contact: {
      authentication_url:
        'https://lion.lyrasistechnology.org/acton/authentication_document',
      contact_email: 'actonlibrary@actonlibrary.org',
      copyright_email: 'actonlibrary@actonlibrary.org',
      help_email: 'actonlibrary@actonlibrary.org',
      opds_url: 'https://lion.lyrasistechnology.org/acton/',
      web_url: 'http://www.actonlibrary.org/',
    },
    areas: {
      focus: ['06475'],
    },
    stages: {
      library_stage: 'production',
      registry_stage: 'production',
    },
  },
  {
    uuid: '2',
    basic_info: {
      name: 'Ak-Chin Indian Community Library, AZ',
      description: 'Serving the needs of library users in Arizona',
      short_name: 'LVYZVC',
      number_of_patrons: '5',
    },
    urls_and_contact: {
      authentication_url:
        'https://simplye.amigos.org/azakchinindiancommunity/authentication_document',
      contact_email: 'simplye@amigos.org',
      copyright_email: 'library@ak-chin.nsn.us',
      help_email: 'library@ak-chin.nsn.us',
      opds_url: 'https://simplye.amigos.org/azakchinindiancommunity/',
      web_url:
        'https://ak-chin.nsn.us/index.php/departments/membership-services',
    },
    areas: {
      focus: ['Maricopa, AZ'],
    },
    stages: {
      library_stage: 'production',
      registry_stage: 'production',
    },
  },
  {
    uuid: '3',
    basic_info: {
      name: 'Alameda County Library',
      description: 'Infinite possibilities',
      short_name: 'CALMDA',
      number_of_patrons: '1839',
    },
    urls_and_contact: {
      authentication_url:
        'http://acl.simplye-ca.org/CALMDA/authentication_document',
      contact_email: 'pmackinnon@califa.org',
      copyright_email: 'simplye@aclibrary.org',
      help_email: 'simplye@aclibrary.org',
      opds_url: 'http://acl.simplye-ca.org/CALMDA/',
      web_url: 'http://aclibrary.org/',
    },
    areas: {
      focus: [],
    },
    stages: {
      library_stage: 'production',
      registry_stage: 'production',
    },
  },
  {
    uuid: '4',
    basic_info: {
      name: 'Alicia Salinas City of Alice Public Library, TX',
      description:
        'Serving the needs of library users in Jim Wells County, Texas',
      short_name: 'JEVRCG',
      number_of_patrons: '2',
    },
    urls_and_contact: {
      authentication_url:
        'https://simplye-tsl1.amigos.org/txaliciasalinascityofalicepublic/authentication_document',
      contact_email: 'simplye@amigos.org',
      copyright_email: 'library@cityofalice.org',
      help_email: 'library@cityofalice.org',
      opds_url:
        'https://simplye-tsl1.amigos.org/txaliciasalinascityofalicepublic/',
      web_url: 'http://www.cityofalice.org/179/Library',
    },
    areas: {
      focus: ['Jim Wells County, TX', 'Alice, TX'],
    },
    stages: {
      library_stage: 'production',
      registry_stage: 'production',
    },
  },
  {
    uuid: '5',
    basic_info: {
      name: 'Allan Shivers Library & Museum, TX',
      description: 'Serving Woodville, TX',
      short_name: 'JEVRCG',
      number_of_patrons: '0',
    },
    urls_and_contact: {
      authentication_url:
        'https://simplye-tsl1.amigos.org/txallanshiverslibrarymuseum/authentication_document',
      contact_email: 'simplye@amigos.org',
      copyright_email: 'ashivers.library@yahoo.com',
      help_email: 'ashivers.library@yahoo.com',
      opds_url: 'https://simplye-tsl1.amigos.org/txallanshiverslibrarymuseum/',
      web_url: 'https://www.allanshiverslibrary.com/',
    },
    areas: {
      focus: ['Jim Wells County, TX', 'Alice, TX'],
    },
    stages: {
      library_stage: 'testing',
      registry_stage: 'testing',
    },
  },
  {
    uuid: '6',
    basic_info: {
      name: 'Allegany County Library System',
      description: 'Serving patrons of Maryland',
      short_name: 'JGSVTQ',
      number_of_patrons: '0',
    },
    urls_and_contact: {
      authentication_url:
        'https://allegany.simplye-md.org/MDALLEGANY/authentication_document',
      contact_email: 'alleganycountylibrary@alleganycountylibrary.info',
      copyright_email: 'alleganycountylibrary@alleganycountylibrary.info',
      help_email: 'alleganycountylibrary@alleganycountylibrary.info',
      opds_url: 'https://allegany.simplye-md.org/MDALLEGANY/',
      web_url: 'http://www.alleganycountylibrary.info/',
    },
    areas: {
      focus: ['Everywhere'],
    },
    stages: {
      library_stage: 'testing',
      registry_stage: 'canceled',
    },
  },
  {
    uuid: '7',
    basic_info: {
      name: 'Allen Memorial Public Library, TX',
      description: 'Open a book, open a mind. Knowledge is power.',
      short_name: 'UOVOCB',
      number_of_patrons: '9',
    },
    urls_and_contact: {
      authentication_url:
        'https://simplye-tsl1.amigos.org/txallenmemorial/authentication_document',
      contact_email: 'simplye@amigos.org',
      copyright_email: 'hawkins.library@yahoo.com',
      help_email: 'hawkins.library@yahoo.com',
      opds_url: 'https://simplye-tsl1.amigos.org/txallenmemorial/',
      web_url: 'https://www.hawkinslibrary.org/',
    },
    areas: {
      focus: ['Hawkins, TX'],
    },
    stages: {
      library_stage: 'production',
      registry_stage: 'production',
    },
  },
  {
    uuid: '8',
    basic_info: {
      name: 'Alpine Public Library, TX',
      description:
        'Building community, encouraging literacy, and promoting lifelong learning',
      short_name: 'NQSJGC',
      number_of_patrons: '59',
    },
    urls_and_contact: {
      authentication_url:
        'https://simplye-tsl1.amigos.org/txalpine/authentication_document',
      contact_email: 'williams@amigos.org',
      copyright_email: 'don@alpinepubliclibrary.org',
      help_email: 'don@alpinepubliclibrary.org',
      opds_url: 'https://simplye-tsl1.amigos.org/txalpine/',
      web_url: 'https://alpinepubliclibrary.org/',
    },
    areas: {
      focus: ['Alpine, TX'],
    },
    stages: {
      library_stage: 'production',
      registry_stage: 'production',
    },
  },
  {
    uuid: '9',
    basic_info: {
      name: 'Amarillo Public Library, TX',
      description: 'Serving the patrons of Amarillo, TX',
      short_name: 'DTENGN',
      number_of_patrons: '78',
    },
    urls_and_contact: {
      authentication_url:
        'https://simplye-tsl1.amigos.org/txamarillopubliclibrary/authentication_document',
      contact_email: 'simplye@amigos.org',
      copyright_email: 'reference@amarillolibrary.org',
      help_email: 'reference@amarillolibrary.org',
      opds_url: 'https://simplye-tsl1.amigos.org/txamarillopubliclibrary/',
      web_url: 'https://www.amarillolibrary.org/',
    },
    areas: {
      focus: ['Amarillo, TX'],
    },
    stages: {
      library_stage: 'production',
      registry_stage: 'production',
    },
  },
  {
    uuid: '10',
    basic_info: {
      name: 'Amigos Demo Library',
      description: 'Unite Libraries for Action and Strength',
      short_name: 'WBDOBX',
      number_of_patrons: '22',
    },
    urls_and_contact: {
      authentication_url:
        'https://circ05.simplye.amigos.org/txamigosls/authentication_document',
      contact_email: 'simplye@amigos.org',
      copyright_email: 'simplye@amigos.org',
      help_email: 'simplye@amigos.org',
      opds_url: 'https://circ05.simplye.amigos.org/txamigosls/',
      web_url: 'https://www.amigos.org',
    },
    areas: {
      focus: ['Dallas, TX'],
    },
    stages: {
      library_stage: 'production',
      registry_stage: 'production',
    },
  },
  {
    uuid: '11',
    basic_info: {
      name: 'Amigos Test Library Two',
      description: 'Unite Libraries for Action and Strength',
      short_name: 'ZKVRQT',
      number_of_patrons: '0',
    },
    urls_and_contact: {
      authentication_url:
        'https://circ06.dev.simplye.amigos.org/txamgtesttwo/authentication_document',
      contact_email: 'williams@amigos.org',
      copyright_email: 'williams@amigos.org',
      help_email: 'simplye@amigos.org',
      opds_url: 'https://circ06.dev.simplye.amigos.org/txamgtesttwo/',
      web_url: 'https://www.amigos.org/simplye',
    },
    areas: {
      focus: ['Dallas, TX'],
    },
    stages: {
      library_stage: 'testing',
      registry_stage: 'testing',
    },
  },
  {
    uuid: '12',
    basic_info: {
      name: 'Anacortes Public Library',
      description: 'Serving Anacortes, WA',
      short_name: 'DCJROE',
      number_of_patrons: '0',
    },
    urls_and_contact: {
      authentication_url:
        'https://wa-shared.lyrasistechnology.org/WA0032/authentication_document',
      contact_email: 'library@cityofanacortes.org',
      copyright_email: 'library@cityofanacortes.org',
      help_email: 'library@cityofanacortes.org',
      opds_url: 'https://wa-shared.lyrasistechnology.org/WA0032/',
      web_url: 'https://www.anacorteswa.gov/220/Library',
    },
    areas: {
      focus: [],
    },
    stages: {
      library_stage: 'testing',
      registry_stage: 'canceled',
    },
  },
  {
    uuid: '13',
    basic_info: {
      name: 'Andrews County Library, TX',
      description: 'Explore the world of books.',
      short_name: 'IDKEWG',
      number_of_patrons: '1',
    },
    urls_and_contact: {
      authentication_url:
        'https://simplye-tsl1.amigos.org/txandrews/authentication_document',
      contact_email: 'simplye@amigos.org',
      copyright_email: 'dbonham@andrews.lib.tx.us',
      help_email: 'dbonham@andrews.lib.tx.us',
      opds_url: 'https://simplye-tsl1.amigos.org/txandrews/',
      web_url: 'https://www.andrews.lib.tx.us/',
    },
    areas: {
      focus: ['Andrews, TX'],
    },
    stages: {
      library_stage: 'production',
      registry_stage: 'production',
    },
  },
  {
    uuid: '14',
    basic_info: {
      name: 'Anne Arundel Co. Public Library',
      description: 'Educate. Enrich. Inspire.',
      short_name: 'MDANNE',
      number_of_patrons: '0',
    },
    urls_and_contact: {
      authentication_url:
        'https://anne-arundel.simplye-md.org/ANNEARUNDEL/authentication_document',
      contact_email: 'askus@aacpl.libanswers.com',
      copyright_email: 'askus@aacpl.libanswers.com',
      help_email: 'askus@aacpl.libanswers.com',
      opds_url: 'https://anne-arundel.simplye-md.org/ANNEARUNDEL/',
      web_url: 'web_urlhttp://www.aacpl.net',
    },
    areas: {
      focus: ['Everywhere'],
    },
    stages: {
      library_stage: 'production',
      registry_stage: 'canceled',
    },
  },
  {
    uuid: '15',
    basic_info: {
      name: 'Ansonia Public Library',
      description: 'Serving Ansonia, CT',
      short_name: 'CTANSO',
      number_of_patrons: '0',
    },
    urls_and_contact: {
      authentication_url:
        'https://bibliomation.lyrasistechnology.org/ansonia/authentication_document',
      contact_email: 'jonathan.green@lyrasis.org',
      copyright_email: 'jonathan.green@lyrasis.org',
      help_email: 'jonathan.green@lyrasis.org',
      opds_url: 'https://bibliomation.lyrasistechnology.org/ansonia/',
      web_url: 'http://www.ansonialibrary.org/',
    },
    areas: {
      focus: ['Everywhere'],
    },
    stages: {
      library_stage: 'canceled',
      registry_stage: 'canceled',
    },
  },
];

export default libraries;
