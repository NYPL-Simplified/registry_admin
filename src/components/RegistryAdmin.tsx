import React, { useContext, useEffect, useState } from 'react';
import Cookies from 'js-cookie';
import {
  SkeletonLoader,
  TemplateAppContainer,
} from '@nypl/design-system-react-components';

import ActionBar from './ActionBar';
import Header from './Header';
import LibrariesList from './LibrariesList';
import LoginForm from './LoginForm';
import { FETCH_LIBRARIES, REFRESH } from '../constants';
import useTokenContext from '../context/tokenContext';
import useLibrariesContext from '../context/librariesContext';
export interface LibraryData {
  areas: {
    focus: string[];
    service: string[];
  };
  basic_info: {
    description: string;
    internal_urn: string;
    name: string;
    number_of_patrons: string;
    online_registration: string;
    pls_id: string;
    short_name: string;
    timestamp: string;
  };
  stages: {
    library_stage: 'testing' | 'production' | 'cancelled';
    registry_stage: 'testing' | 'production' | 'cancelled';
  };
  urls_and_contact: {
    authentication_url: string;
    contact_email: string | null;
    contact_validated: string;
    copyright_email: string | null;
    copyright_validated: string;
    help_email: string | null;
    help_validated: string;
    opds_url: string;
    web_url: string;
  };
  uuid: string;
}

const RegistryAdmin = () => {
  // If isLoading is true, a SkeletonLoader is displayed. isLoading is set to
  // false after fetching and/or accessToken refreshing is completed.
  const [isLoading, setIsLoading] = useState<boolean>(true);
  // isSimpleList determines whether the libraries will be rendered as accordions
  // with full details, or as a table of names and patron counts.
  const [isSimpleList, setIsSimpleList] = useState<boolean>(false);

  const { accessToken, setAccessToken } = useTokenContext();
  const { setLibraries } = useLibrariesContext();

  // The logout function resets the accessToken to an empty string and deletes
  // the refreshToken from cookie storage. This causes the LoginForm to be
  // displayed.
  const logout = () => {
    setAccessToken('');
    Cookies.remove('refreshToken');
    setIsLoading(false);
  };

  const refresh = () => {
    // Look for the refreshToken cookie.
    const refreshToken = Cookies.get('refreshToken');

    // If a refreshToken exists, send it as a header to the refresh endpoint.
    if (refreshToken) {
      fetch(REFRESH as RequestInfo | URL, {
        method: 'POST',
        headers: { Authorization: `Bearer ${refreshToken}` },
      })
        .then((response) => {
          if (response.ok) {
            return response.json();
          } else {
            throw new Error('There was a problem refreshing.');
          }
        })
        // If the call is successful, save the new accessToken to tokenContext.
        .then((data) => {
          setAccessToken(data.access_token);
          setIsLoading(false);
        })
        // If there is a problem refreshing, log the error message and use the
        // logout function to delete the accessToken (if one exists) and the
        // refreshToken. This will bring the user back to the LoginForm.
        .catch((err) => {
          console.log(err.message);
          logout();
        });
    } else {
      // If no refreshToken exists, delete the accessToken (if there is one)
      // so the user is brought back to the LoginForm.
      logout();
    }
  };

  // Fetch the libraries with the accessToken.
  const fetchLibraries = () => {
    fetch(FETCH_LIBRARIES as RequestInfo | URL, {
      headers: { Authorization: `Bearer ${accessToken}` },
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
          // If the response status is 401, the accessToken may be expired.
          // Try refreshing it.
        } else if (response.status === 401) {
          refresh();
          return;
        } else {
          throw new Error('There was a problem fetching libraries.');
        }
      })
      .then((data) => {
        setLibraries(data.libraries);
        setIsLoading(false);
      })
      .catch((err) => {
        console.log(err.message);
        logout();
      });
  };

  // If there is no accessToken, try using the refresh endpoint to get a new
  // one. If there is, fetch the libraries.
  useEffect(() => {
    if (!accessToken) {
      refresh();
    } else {
      fetchLibraries();
    }
  }, [accessToken]);

  return (
    <>
      {isLoading ? (
        <SkeletonLoader />
      ) : (
        <TemplateAppContainer
          header={<Header />}
          contentTop={
            accessToken ? (
              <>
                <ActionBar
                  logout={logout}
                  isSimpleList={isSimpleList}
                  setIsSimpleList={setIsSimpleList}
                />
              </>
            ) : undefined
          }
          contentPrimary={
            accessToken ? (
              <LibrariesList isSimpleList={isSimpleList} />
            ) : (
              <LoginForm />
            )
          }
        />
      )}
    </>
  );
};

export default RegistryAdmin;
