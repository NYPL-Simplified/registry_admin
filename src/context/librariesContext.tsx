import React, {
  createContext,
  Dispatch,
  useContext,
  useEffect,
  useState,
} from 'react';
export interface LibraryData {
  areas: {
    focus: string[];
    service: string[];
  };
  basic_info: {
    description: string;
    internal_urn: string;
    name: string;
    number_of_patrons: string;
    online_registration: string;
    pls_id: string;
    short_name: string;
    timestamp: string;
  };
  stages: {
    library_stage: 'testing' | 'production' | 'cancelled';
    registry_stage: 'testing' | 'production' | 'cancelled';
  };
  urls_and_contact: {
    authentication_url: string;
    contact_email: string | null;
    contact_validated: string;
    copyright_email: string | null;
    copyright_validated: string;
    help_email: string | null;
    help_validated: string;
    opds_url: string;
    web_url: string;
  };
  uuid: string;
}

type LibrariesContextValues = {
  libraries: LibraryData[];
  setLibraries: Dispatch<React.SetStateAction<LibraryData[]>>;
  setUpdatedLibrary: Dispatch<React.SetStateAction<LibraryData | undefined>>;
};

type LibrariesProviderProps = {
  children: React.ReactNode;
};

export const LibrariesContext = createContext<
  LibrariesContextValues | undefined
>(undefined);

export const LibrariesProvider = ({ children }: LibrariesProviderProps) => {
  const [updatedLibrary, setUpdatedLibrary] = useState<
    LibraryData | undefined
  >();
  const [libraries, setLibraries] = useState<LibraryData[]>([]);

  useEffect(() => {
    if (updatedLibrary) {
      // Find the index of the library with the same uuid as the one
      // that was recently updated.
      const indexOfElementToReplace = libraries.findIndex(
        (library) => library.uuid === updatedLibrary.uuid
      );

      // Replace the library at that index with the updatedLibrary.
      const updatedLibraries = libraries.slice();
      updatedLibraries.splice(indexOfElementToReplace, 1, updatedLibrary);

      // Set libraries to the updated list.
      setLibraries(updatedLibraries);
    }
  }, [updatedLibrary]);

  return (
    <LibrariesContext.Provider
      value={{ libraries, setLibraries, setUpdatedLibrary }}
    >
      {children}
    </LibrariesContext.Provider>
  );
};

const useLibrariesContext = () => {
  const context = useContext(LibrariesContext);
  if (typeof context === 'undefined') {
    throw new Error('useLibraries must be used within a LibraryProvider');
  }
  return context;
};

export default useLibrariesContext;
