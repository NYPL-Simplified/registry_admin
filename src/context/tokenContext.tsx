import React, { createContext, Dispatch, useState } from 'react';

type TokenContextValues = {
  accessToken: string;
  setAccessToken: Dispatch<React.SetStateAction<string>>;
};

type TokenProviderProps = {
  children: React.ReactNode;
};

export const TokenContext = createContext<TokenContextValues | undefined>(
  undefined
);

export const TokenProvider = ({ children }: TokenProviderProps) => {
  const [accessToken, setAccessToken] = useState<string>('');

  return (
    <TokenContext.Provider value={{ accessToken, setAccessToken }}>
      {children}
    </TokenContext.Provider>
  );
};

const useTokenContext = () => {
  const context = React.useContext(TokenContext);
  if (typeof context === 'undefined') {
    throw new Error('useLibraries must be used within a LibraryProvider');
  }
  return context;
};

export default useTokenContext;
