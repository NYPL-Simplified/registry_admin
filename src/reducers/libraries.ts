import { LibrariesData } from "../interfaces";
import ActionCreator from "../actions";
import createFetchEditReducer from "./createFetchEditReducer";

export const libraries = createFetchEditReducer<LibrariesData>(ActionCreator.GET_ALL_LIBRARIES);
export const results = createFetchEditReducer<LibrariesData>(ActionCreator.GET_SOME_LIBRARIES);
