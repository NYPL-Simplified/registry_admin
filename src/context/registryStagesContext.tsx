import React, { createContext, Dispatch, useState } from 'react';
import { LibraryStage } from '../components/LibraryDetails';

export type RegistryStagesContextValues = {
  registryStages: Record<string, LibraryStage>;
  setRegistryStages: Dispatch<
    React.SetStateAction<Record<string, LibraryStage>>
  >;
};

type RegistryStagesProviderProps = {
  children: React.ReactNode;
};

export const RegistryStagesContext =
  createContext<RegistryStagesContextValues | null>(null);

export const RegistryStagesProvider = ({
  children,
}: RegistryStagesProviderProps) => {
  const [registryStages, setRegistryStages] = useState<
    Record<string, LibraryStage>
  >({});

  console.log('registryStages -->', registryStages);

  return (
    <RegistryStagesContext.Provider
      value={{ registryStages, setRegistryStages }}
    >
      {children}
    </RegistryStagesContext.Provider>
  );
};
