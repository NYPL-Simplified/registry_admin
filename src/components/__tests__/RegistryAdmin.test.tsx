import * as React from 'react';
import Cookies from 'js-cookie';
import { act, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import libraries from '../../../data/libraries';
import RegistryAdmin from '../RegistryAdmin';
import { TokenContext } from '../../context/tokenContext';

const setAccessTokenMock = jest.fn();

const renderRegistryAdminWithContext = (
  accessToken: string,
  setAccessToken = setAccessTokenMock
) => {
  render(
    <TokenContext.Provider value={{ accessToken, setAccessToken }}>
      <RegistryAdmin />
    </TokenContext.Provider>
  );
};

const renderRegistryForSnapshot = (
  accessToken: string,
  setAccessToken = setAccessTokenMock
) => {
  return (
    <TokenContext.Provider value={{ accessToken, setAccessToken }}>
      <RegistryAdmin />
    </TokenContext.Provider>
  );
};

describe('RegistryAdmin, with no access token', () => {
  beforeEach(() => {
    renderRegistryAdminWithContext('');
  });

  it('renders the header', () => {
    const header = screen.getByRole('heading', { level: 1 });

    expect(header).toHaveTextContent(/library registry interface/i);
  });

  it('renders the login form', () => {
    const usernameInput = screen.getByRole('textbox');
    const passwordInput = screen.getByLabelText(/password/i);

    expect(usernameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();

    const submitButton = screen.getByRole('button');

    expect(submitButton).toHaveTextContent(/submit/i);
  });

  it('does not render admin page elements', () => {
    const listViewToggle = screen.queryByRole('checkbox');
    const buttons = screen.getAllByRole('button');

    expect(listViewToggle).not.toBeInTheDocument();
    // There are 16 buttons on the admin page, but only 1 on the
    // login page.
    expect(buttons.length).toEqual(1);
  });
});

describe('RegistryAdmin, with access token', () => {
  beforeEach(async () => {
    (global as any).fetch = jest.fn().mockReturnValueOnce(
      Promise.resolve({
        ok: true,
        status: 200,
        json: () => Promise.resolve({ libraries: libraries }),
      })
    );

    await act(() => renderRegistryAdminWithContext('mockAccessToken'));
  });

  it('renders the header', () => {
    const header = screen.getByRole('heading', { level: 1 });

    expect(header).toHaveTextContent(/library registry interface/i);
  });

  it('renders the list toggle and logout button', () => {
    const listViewToggle = screen.getByRole('checkbox');
    const logoutButton = screen.getAllByRole('button')[0];

    expect(listViewToggle).toBeInTheDocument();
    expect(logoutButton).toHaveTextContent(/log out/i);
  });

  it('renders the list of libraries', () => {
    const libraryAccordions = screen.getAllByRole('button').slice(1);

    expect(libraryAccordions.length).toEqual(15);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});

describe('RegistryAdmin, with no access token, but with refresh token', () => {
  it('tries to refresh the accessToken', async () => {
    const realGet = Cookies.get;

    Cookies.get = jest.fn().mockImplementation(() => 'mockRefreshToken');
    (global as any).fetch = jest.fn().mockReturnValueOnce(
      Promise.resolve({
        ok: true,
        status: 200,
        json: () => Promise.resolve({ access_token: 'mockAccessToken' }),
      })
    );

    await act(() => renderRegistryAdminWithContext(''));

    expect(fetch).toHaveBeenCalledWith(process.env.REFRESH, {
      headers: { Authorization: 'Bearer mockRefreshToken' },
      method: 'POST',
    });

    Cookies.get = realGet;
    jest.clearAllMocks();
  });
});

describe('RegistryAdmin Snapshot', () => {
  it('renders the UI snapshot correctly', async () => {
    (global as any).fetch = jest.fn().mockReturnValueOnce(
      Promise.resolve({
        ok: true,
        status: 200,
        json: () => Promise.resolve({ libraries: libraries }),
      })
    );
    let loginPage;
    let adminPage;
    await act(() => {
      loginPage = renderer.create(renderRegistryForSnapshot('')).toJSON();
      expect(loginPage).toMatchSnapshot();
    });

    await act(() => {
      adminPage = renderer
        .create(renderRegistryForSnapshot('mockAccessToken'))
        .toJSON();
      expect(adminPage).toMatchSnapshot();
    });

    jest.clearAllMocks();
  });
});
