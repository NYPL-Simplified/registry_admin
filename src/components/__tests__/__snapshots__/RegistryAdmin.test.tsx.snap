// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistryAdmin Snapshot renders the UI snapshot correctly 1`] = `null`;

exports[`RegistryAdmin Snapshot renders the UI snapshot correctly 2`] = `null`;
